File: #0108.convert-sorted-array-to-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* sortedArrayToBST(vector<int>& nums) {

        function<TreeNode*(int left, int right)> binarySearchTree = [&](int left, int right) {
            if (left > right) {
                return static_cast<TreeNode*>(nullptr);
            }
            int middle = left + (right - left) / 2;
            TreeNode* root = new TreeNode(nums[middle]);
            root->left = binarySearchTree(left, middle - 1);
            root->right = binarySearchTree(middle + 1, right);
            return root;
        };
        return binarySearchTree(0, nums.size() - 1);
    }
};

File: #0108.convert-sorted-array-to-binary-search-tree.rs
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
// 
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::rc::Rc;
use std::cell::RefCell;
impl Solution {
    pub fn sorted_array_to_bst(nums: Vec<i32>) -> Option<Rc<RefCell<TreeNode>>> {
        return Self::binary_search_tree(&nums, 0, nums.len());
    }

    fn binary_search_tree(nums: &Vec<i32>, left: usize, right: usize) -> Option<Rc<RefCell<TreeNode>>> {
        if left >= right {
            None
        } else {
            let middle = left + (right - left) / 2;
            Some(Rc::new(RefCell::new(TreeNode {
                val: nums[middle],
                right: Self::binary_search_tree(nums, middle + 1, right),
                left: Self::binary_search_tree(nums, left, middle)
            })))
        }
    }
}
